src/sage/doctest/parsing.py

```diff
@@ -80,111 +94,448 @@ def fake_RIFtol(*args):
 ansi_escape_sequence = re.compile(r'(\x1b[@-Z\\-~]|\x1b\[.*?[@-~]|\x9b.*?[@-~])')
 
 special_optional_regex = 'arb216|arb218|py2|long time|not implemented|not tested|known bug'
-optional_regex = re.compile(fr'({special_optional_regex})|([^ a-z]\s*optional\s*[:-]*((\s|\w|[.])*))')
-special_optional_regex = re.compile(special_optional_regex)
+tag_with_explanation_regex = fr'((?:\w|[.])+)\s*(?:\((.*?)\))?'
+optional_regex = re.compile(fr'(?P<cmd>{special_optional_regex})\s*(?:\((?P<cmd_explanation>.*?)\))?|'
+                            fr'[^ a-z]\s*(optional|needs)(?:\s|[:-])*(?P<tags>(?:(?:{tag_with_explanation_regex})\s*)*)',
+                            re.IGNORECASE)
+special_optional_regex = re.compile(special_optional_regex, re.IGNORECASE)
+tag_with_explanation_regex = re.compile(tag_with_explanation_regex, re.IGNORECASE)
 
+nodoctest_regex = re.compile(r'\s*(#+|%+|r"+|"+|\.\.)\s*nodoctest')
+optionaltag_regex = re.compile(r'^(\w|[.])+$')
+optionalfiledirective_regex = re.compile(r'\s*(#+|%+|r"+|"+|\.\.)\s*sage\.doctest: (.*)')
 
-def parse_optional_tags(string):
-    """
-    Return a set consisting of the optional tags from the following
+
+def parse_optional_tags(string, *, return_string_sans_tags=False):
+    r"""
+    Return a dictionary whose keys are optional tags from the following
     set that occur in a comment on the first line of the input string.
 
-    - 'long time'
-    - 'not implemented'
-    - 'not tested'
-    - 'known bug'
-    - 'py2'
-    - 'arb216'
-    - 'arb218'
-    - 'optional: PKG_NAME' -- the set will just contain 'PKG_NAME'
+    - ``'long time'``
+    - ``'not implemented'``
+    - ``'not tested'``
+    - ``'known bug'``
+    - ``'py2'``
+    - ``'arb216'``
+    - ``'arb218'``
+    - ``'optional - PKG_NAME...'`` -- the dictionary will just have the key ``'PKG_NAME'``
+
```

Comment:
Perhaps `needs PKG_NAME...` should be in the list.

Reply:
done in 44e5752545; I've changed from PKG_NAME to FEATURE